File: experiment/backends/android/dazzle/DazzleLuaContextManager.kt
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package dev.dazzle.experiment

import dev.dazzle.sdk.Dazzle
import dev.dazzle.sdk.DazzleServer
import dev.dazzle.sdk.StreamKey

/**
 * Optimized Valkey [StorageBackend] that uses Lua scripting + directPipeline
 * to demonstrate that Valkey's capabilities not only provide richer primitives
 * but also **faster retrieval** than any other backend when used correctly.
 *
 * The naive ValkeyContextManager issues 6-8 separate directCommand calls per
 * buildContextBlock (XREVRANGE + 5 HGETs + ZRANGEBYSCORE), each crossing the
 * event-loop pipe independently → ~4.4 ms total.
 *
 * This optimized version:
 *   1. **Writes** use directPipeline to batch all ingestion ops into ONE
 *      pipe dispatch (1 round-trip instead of 6-8).
 *   2. **Reads** use a single EVAL Lua script that executes all queries
 *      server-side and returns the assembled context block as a single
 *      string (1 round-trip instead of 6-8).
 *
 * Expected improvement: retrieval drops from ~4.4 ms to ~500-800 µs,
 * making Valkey FASTER than SQLite (1.4 ms) while retaining the full
 * capability set (HFE, HLL, Geo, Pub/Sub, Streams, etc.).
 *
 * This is the paper's key demonstration: "Valkey's scripting capability
 * is not just a feature checkbox — it directly enables a 5-8× retrieval
 * speedup that simpler backends cannot achieve."
 */
class DazzleLuaContextManager : StorageBackend {

    override val backendName: String = "Dazzle-Lua"

    private val dazzle: Dazzle = DazzleServer.client()

    // Keep the typed primitives for writes and for the non-Lua fallbacks
    private val readings  = dazzle.stream("sensor:readings")
    private val stats     = dazzle.hash("sensor:stats")
    private val anomalies = dazzle.sortedSet("sensor:anomalies")
    private val decisions = dazzle.list("agent:decisions")

    // ── Lua scripts (loaded once, cached by SHA) ──────────────────────────

    // Lua script that fetches ALL data in ONE round-trip and returns raw
    // values as a flat array. Kotlin formats the output (handling Unicode
    // chars like ° and → that break RESP byte-count parsing).
    private val contextBlockScript = dazzle.script("""
        local result = {}
        local entries = redis.call('XREVRANGE', KEYS[1], '+', '-', 'COUNT', 10)
        local temps = {}
        for i = #entries, 1, -1 do
            local fields = entries[i][2]
            for j = 1, #fields, 2 do
                if fields[j] == 'temp' then temps[#temps+1] = fields[j+1] end
            end
        end
        result[1] = table.concat(temps, ',')
        local s = redis.call('HMGET', KEYS[2], 'count', 'temp_sum', 'min_temp', 'max_temp', 'anomaly_count')
        result[2] = s[1] or '0'
        result[3] = s[2] or '0'
        result[4] = s[3] or '0'
        result[5] = s[4] or '0'
        result[6] = s[5] or '0'
        local anoms = redis.call('ZRANGEBYSCORE', KEYS[3], ARGV[1], ARGV[2])
        result[7] = table.concat(anoms, ',')
        return result
    """.trimIndent())

    private val synthesisScript = dazzle.script("""
        local result = {}
        local s = redis.call('HMGET', KEYS[1], 'count', 'temp_sum', 'min_temp', 'max_temp', 'anomaly_count')
        result[1] = s[1] or '0'
        result[2] = s[2] or '0'
        result[3] = s[3] or '0'
        result[4] = s[4] or '0'
        result[5] = s[5] or '0'
        local anoms = redis.call('ZRANGEBYSCORE', KEYS[2], '-inf', '+inf')
        result[6] = table.concat(anoms, ',')
        local decs = redis.call('LRANGE', KEYS[3], 0, -1)
        result[7] = table.concat(decs, '|')
        return result
    """.trimIndent())

    // ── Lifecycle ─────────────────────────────────────────────────────────

    override fun flush() {
        readings.deleteKey()
        stats.delete()
        anomalies.deleteKey()
        decisions.deleteKey()
        for (i in 0..9) dazzle.hash("agent:checkpoint:$i").delete()
    }

    // ── Ingest (batched via directPipeline) ───────────────────────────────

    override fun ingest(reading: SensorReading) {
        // Batch all writes into one pipeline dispatch
        val commands = mutableListOf<List<String>>()

        commands.add(listOf(
            "XADD", "sensor:readings", "MAXLEN", "~", "200", "*",
            "temp", reading.tempC.toString(),
            "humidity", reading.humidity.toString(),
            "minute", reading.minute.toString(),
            "anomalous", if (reading.anomalous) "1" else "0",
        ))
        commands.add(listOf("HINCRBYFLOAT", "sensor:stats", "temp_sum", reading.tempC.toString()))
        commands.add(listOf("HINCRBY", "sensor:stats", "count", "1"))
        commands.add(listOf("HSET", "sensor:stats", "latest_temp", reading.tempC.toString()))
        commands.add(listOf("HSET", "sensor:stats", "latest_minute", reading.minute.toString()))

        if (reading.anomalous) {
            commands.add(listOf("ZADD", "sensor:anomalies", reading.minute.toString(), reading.minute.toString()))
            commands.add(listOf("HINCRBY", "sensor:stats", "anomaly_count", "1"))
        }

        DazzleServer.directPipeline(commands)

        // Min/max still needs read-then-write (can't pipeline conditional logic)
        val curMin = stats.get("min_temp")?.toDoubleOrNull()
        if (curMin == null || reading.tempC < curMin) {
            stats.set("min_temp", reading.tempC.toString())
        }
        val curMax = stats.get("max_temp")?.toDoubleOrNull()
        if (curMax == null || reading.tempC > curMax) {
            stats.set("max_temp", reading.tempC.toString())
        }
    }

    // ── Context block (1 Lua EVAL = 1 pipe round-trip) ────────────────────

    override fun buildContextBlock(currentMinute: Int, windowMinutes: Int): String {
        val windowStart = maxOf(0, currentMinute - windowMinutes)
        val raw = contextBlockScript.eval(
            keys = listOf("sensor:readings", "sensor:stats", "sensor:anomalies"),
            args = listOf(windowStart.toString(), currentMinute.toString(), windowMinutes.toString()),
        ).asArray().map { it.asBulkOrNull() ?: "" }

        // raw[0] = temps CSV, raw[1..5] = count/sum/min/max/anomCnt, raw[6] = anomaly mins CSV
        val lines = mutableListOf<String>()

        val tempsCsv = raw.getOrNull(0) ?: ""
        if (tempsCsv.isNotEmpty()) {
            val temps = tempsCsv.split(",").mapNotNull { it.toDoubleOrNull() }
            if (temps.isNotEmpty()) {
                val formatted = temps.joinToString(", ") { String.format("%.1f", it) }
                lines += "Last ${temps.size} temperatures (oldest→newest, °C): $formatted"
                lines += "Recent trend: ${computeTrend(temps)}"
            }
        }

        val count = raw.getOrNull(1)?.toIntOrNull() ?: 0
        if (count > 0) {
            val sum = raw.getOrNull(2)?.toDoubleOrNull() ?: 0.0
            val minT = raw.getOrNull(3)?.toDoubleOrNull() ?: 0.0
            val maxT = raw.getOrNull(4)?.toDoubleOrNull() ?: 0.0
            val anomCnt = raw.getOrNull(5)?.toIntOrNull() ?: 0
            lines += "Aggregate over $count readings: " +
                "avg=${String.format("%.1f", sum / count)}°C, " +
                "min=${String.format("%.1f", minT)}°C, " +
                "max=${String.format("%.1f", maxT)}°C"
            lines += "Total anomalies detected so far: $anomCnt"
        }

        val anomsCsv = raw.getOrNull(6) ?: ""
        if (anomsCsv.isEmpty()) {
            lines += "No anomalies in the last $windowMinutes minutes."
        } else {
            lines += "Anomalous time indices in the last $windowMinutes minutes " +
                "(minute numbers, not temperatures): [$anomsCsv]"
        }

        return lines.joinToString("\n")
    }

    override fun buildSynthesisContext(): String {
        val raw = synthesisScript.eval(
            keys = listOf("sensor:stats", "sensor:anomalies", "agent:decisions"),
            args = emptyList(),
        ).asArray().map { it.asBulkOrNull() ?: "" }

        // raw[0..4] = count/sum/min/max/anomCnt, raw[5] = anomaly mins CSV, raw[6] = decisions pipe-separated
        val lines = mutableListOf<String>()

        val count = raw.getOrNull(0)?.toIntOrNull() ?: 0
        if (count > 0) {
            val sum = raw.getOrNull(1)?.toDoubleOrNull() ?: 0.0
            val minT = raw.getOrNull(2)?.toDoubleOrNull() ?: 0.0
            val maxT = raw.getOrNull(3)?.toDoubleOrNull() ?: 0.0
            val anomCnt = raw.getOrNull(4)?.toIntOrNull() ?: 0
            lines += "=== Full Session Stats ==="
            lines += "Total readings: $count"
            lines += "Temperature range: ${String.format("%.1f", minT)}°C to " +
                "${String.format("%.1f", maxT)}°C " +
                "(avg ${String.format("%.1f", sum / count)}°C)"
            lines += "Total anomalies detected: $anomCnt"
        }

        val anomsCsv = raw.getOrNull(5) ?: ""
        if (anomsCsv.isNotEmpty()) {
            lines += "Anomalous time indices (minute numbers, not temperatures): [$anomsCsv]"
        }

        val decsPipe = raw.getOrNull(6) ?: ""
        if (decsPipe.isNotEmpty()) {
            val decs = decsPipe.split("|")
            lines += "=== Monitoring Agent Decisions ==="
            for ((idx, decision) in decs.withIndex()) {
                lines += "  Checkpoint ${idx + 1}: $decision"
            }
        }

        return lines.joinToString("\n")
    }

    // ── Decision storage ──────────────────────────────────────────────────

    private fun computeTrend(temps: List<Double>): String {
        if (temps.size < 2) return "stable"
        val n = temps.size
        val meanX = (n - 1) / 2.0
        val meanY = temps.average()
        val num = (0 until n).sumOf { (it - meanX) * (temps[it] - meanY) }
        val den = (0 until n).sumOf { (it - meanX) * (it - meanX) }
        val slope = if (den != 0.0) num / den else 0.0
        return when {
            slope >  0.15 -> "increasing"
            slope < -0.15 -> "decreasing"
            else          -> "stable"
        }
    }

    override fun storeCheckpointDecision(
        index: Int,
        minute: Int,
        anomalyDetected: Boolean,
        severity: String,
        trend: String,
    ) {
        val decision = "anomaly=${if (anomalyDetected) "yes" else "no"} " +
            "severity=$severity trend=$trend"

        DazzleServer.directPipeline(listOf(
            listOf("HSET", "agent:checkpoint:$index",
                "minute", minute.toString(),
                "anomaly", if (anomalyDetected) "1" else "0",
                "severity", severity,
                "trend", trend),
            listOf("RPUSH", "agent:decisions", decision),
        ))
    }

    // ── Footprint accounting ──────────────────────────────────────────────

    override val backendSizeMethod: String = "valkey:used_memory_dataset"

    override fun backendSizeBytes(): Long {
        val info = dazzle.server().info("memory")
        return info.field("Memory", "used_memory_dataset")?.toLongOrNull()
            ?: info.usedMemoryBytes
            ?: -1L
    }

    override fun backendSizeBreakdown(): Map<String, Long> =
        valkeyMemoryBreakdown(dazzle.server().info("memory"))
}

File: experiment/backends/ios/dazzle/DazzleLuaContextManager.swift
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation

/// Optimized Valkey StorageBackend that uses Lua scripting to demonstrate
/// that Valkey's capabilities enable faster retrieval than any other
/// backend when used correctly.
///
/// Reads use a single EVAL Lua script that executes all queries
/// server-side and returns the assembled data as a single array
/// (1 round-trip instead of 6-8).
final class DazzleLuaContextManager: StorageBackend {

    let backendName: String = "Dazzle-Lua"

    private let dazzle: Dazzle
    private let readings:  StreamKey
    private let stats:     HashKey
    private let anomalies: SortedSetKey
    private let decisions: ListKey

    // Lua scripts (loaded once, cached by SHA)
    private let contextBlockScript: LuaScript
    private let synthesisScript:    LuaScript

    init() {
        let client = DazzleServer.shared.client()
        self.dazzle    = client
        self.readings  = client.stream("sensor:readings")
        self.stats     = client.hash("sensor:stats")
        self.anomalies = client.sortedSet("sensor:anomalies")
        self.decisions = client.list("agent:decisions")

        // Fetches ALL data in ONE round-trip, returns raw values as flat array.
        // Swift formats the output (handling Unicode chars like degree and arrow).
        self.contextBlockScript = client.script("""
            local result = {}
            local entries = redis.call('XREVRANGE', KEYS[1], '+', '-', 'COUNT', 10)
            local temps = {}
            for i = #entries, 1, -1 do
                local fields = entries[i][2]
                for j = 1, #fields, 2 do
                    if fields[j] == 'temp' then temps[#temps+1] = fields[j+1] end
                end
            end
            result[1] = table.concat(temps, ',')
            local s = redis.call('HMGET', KEYS[2], 'count', 'temp_sum', 'min_temp', 'max_temp', 'anomaly_count')
            result[2] = s[1] or '0'
            result[3] = s[2] or '0'
            result[4] = s[3] or '0'
            result[5] = s[4] or '0'
            result[6] = s[5] or '0'
            local anoms = redis.call('ZRANGEBYSCORE', KEYS[3], ARGV[1], ARGV[2])
            result[7] = table.concat(anoms, ',')
            return result
            """)

        self.synthesisScript = client.script("""
            local result = {}
            local s = redis.call('HMGET', KEYS[1], 'count', 'temp_sum', 'min_temp', 'max_temp', 'anomaly_count')
            result[1] = s[1] or '0'
            result[2] = s[2] or '0'
            result[3] = s[3] or '0'
            result[4] = s[4] or '0'
            result[5] = s[5] or '0'
            local anoms = redis.call('ZRANGEBYSCORE', KEYS[2], '-inf', '+inf')
            result[6] = table.concat(anoms, ',')
            local decs = redis.call('LRANGE', KEYS[3], 0, -1)
            result[7] = table.concat(decs, '|')
            return result
            """)
    }

    func flush() {
        _ = try? readings.deleteKey()
        _ = try? stats.deleteKey()
        _ = try? anomalies.deleteKey()
        _ = try? decisions.deleteKey()
        for i in 0...9 { _ = try? dazzle.hash("agent:checkpoint:\(i)").deleteKey() }
    }

    // MARK: - Ingest (standard typed primitives)

    func ingest(_ reading: SensorReading) {
        _ = try? readings.add(
            fields: [
                ("temp",      String(reading.tempC)),
                ("humidity",  String(reading.humidity)),
                ("minute",    String(reading.minute)),
                ("anomalous", reading.anomalous ? "1" : "0"),
            ],
            maxLen: 200
        )
        _ = try? stats.incrByFloat("temp_sum", reading.tempC)
        _ = try? stats.incrBy("count", 1)
        _ = try? stats.set("latest_temp", String(reading.tempC))
        _ = try? stats.set("latest_minute", String(reading.minute))

        if reading.anomalous {
            _ = try? anomalies.add(score: Double(reading.minute), member: String(reading.minute))
            _ = try? stats.incrBy("anomaly_count", 1)
        }

        // Min/max (conditional)
        let curMin = (try? stats.get("min_temp")).flatMap { $0 }.flatMap { Double($0) }
        if curMin == nil || reading.tempC < curMin! {
            _ = try? stats.set("min_temp", String(reading.tempC))
        }
        let curMax = (try? stats.get("max_temp")).flatMap { $0 }.flatMap { Double($0) }
        if curMax == nil || reading.tempC > curMax! {
            _ = try? stats.set("max_temp", String(reading.tempC))
        }
    }

    // MARK: - Context block (1 Lua EVAL = 1 round-trip)

    func buildContextBlock(currentMinute: Int, windowMinutes: Int = 20) -> String {
        let windowStart = max(0, currentMinute - windowMinutes)
        guard let reply = try? contextBlockScript.eval(
            keys: ["sensor:readings", "sensor:stats", "sensor:anomalies"],
            args: [String(windowStart), String(currentMinute), String(windowMinutes)]
        ) else { return "" }

        let raw = reply.asArray.map { $0.asBulkOrNil ?? "" }
        var lines: [String] = []

        // raw[0] = temps CSV
        let tempsCsv = raw.first ?? ""
        if !tempsCsv.isEmpty {
            let temps = tempsCsv.split(separator: ",").compactMap { Double($0) }
            if !temps.isEmpty {
                let formatted = temps.map { String(format: "%.1f", $0) }.joined(separator: ", ")
                lines.append("Last \(temps.count) temperatures (oldest\u{2192}newest, \u{00B0}C): \(formatted)")
                lines.append("Recent trend: \(computeTrend(temps))")
            }
        }

        // raw[1..5] = count/sum/min/max/anomCnt
        if let count = Int(raw.count > 1 ? raw[1] : "0"), count > 0 {
            let sum  = Double(raw.count > 2 ? raw[2] : "0") ?? 0
            let minT = Double(raw.count > 3 ? raw[3] : "0") ?? 0
            let maxT = Double(raw.count > 4 ? raw[4] : "0") ?? 0
            let anomCnt = Int(raw.count > 5 ? raw[5] : "0") ?? 0
            lines.append("Aggregate over \(count) readings: " +
                "avg=\(String(format: "%.1f", sum / Double(count)))\u{00B0}C, " +
                "min=\(String(format: "%.1f", minT))\u{00B0}C, " +
                "max=\(String(format: "%.1f", maxT))\u{00B0}C")
            lines.append("Total anomalies detected so far: \(anomCnt)")
        }

        // raw[6] = anomaly mins CSV
        let anomsCsv = raw.count > 6 ? raw[6] : ""
        if anomsCsv.isEmpty {
            lines.append("No anomalies in the last \(windowMinutes) minutes.")
        } else {
            lines.append("Anomalous time indices in the last \(windowMinutes) minutes " +
                "(minute numbers, not temperatures): [\(anomsCsv)]")
        }

        return lines.joined(separator: "\n")
    }

    func buildSynthesisContext() -> String {
        guard let reply = try? synthesisScript.eval(
            keys: ["sensor:stats", "sensor:anomalies", "agent:decisions"],
            args: []
        ) else { return "" }

        let raw = reply.asArray.map { $0.asBulkOrNil ?? "" }
        var lines: [String] = []

        // raw[0..4] = count/sum/min/max/anomCnt
        if let count = Int(raw.count > 0 ? raw[0] : "0"), count > 0 {
            let sum  = Double(raw.count > 1 ? raw[1] : "0") ?? 0
            let minT = Double(raw.count > 2 ? raw[2] : "0") ?? 0
            let maxT = Double(raw.count > 3 ? raw[3] : "0") ?? 0
            let anomCnt = Int(raw.count > 4 ? raw[4] : "0") ?? 0
            lines.append("=== Full Session Stats ===")
            lines.append("Total readings: \(count)")
            lines.append("Temperature range: \(String(format: "%.1f", minT))\u{00B0}C to " +
                "\(String(format: "%.1f", maxT))\u{00B0}C " +
                "(avg \(String(format: "%.1f", sum / Double(count)))\u{00B0}C)")
            lines.append("Total anomalies detected: \(anomCnt)")
        }

        // raw[5] = anomaly mins CSV
        let anomsCsv = raw.count > 5 ? raw[5] : ""
        if !anomsCsv.isEmpty {
            lines.append("Anomalous time indices (minute numbers, not temperatures): [\(anomsCsv)]")
        }

        // raw[6] = decisions pipe-separated
        let decsPipe = raw.count > 6 ? raw[6] : ""
        if !decsPipe.isEmpty {
            let decs = decsPipe.split(separator: "|").map(String.init)
            lines.append("=== Monitoring Agent Decisions ===")
            for (idx, decision) in decs.enumerated() {
                lines.append("  Checkpoint \(idx + 1): \(decision)")
            }
        }

        return lines.joined(separator: "\n")
    }

    func storeCheckpointDecision(index: Int, minute: Int, anomalyDetected: Bool, severity: String, trend: String) {
        let decision = "anomaly=\(anomalyDetected ? "yes" : "no") severity=\(severity) trend=\(trend)"
        _ = try? dazzle.hash("agent:checkpoint:\(index)").setAll([
            "minute":   String(minute),
            "anomaly":  anomalyDetected ? "1" : "0",
            "severity": severity,
            "trend":    trend,
        ])
        _ = try? decisions.rpush(decision)
    }

    private func computeTrend(_ temps: [Double]) -> String {
        guard temps.count >= 2 else { return "stable" }
        let n = temps.count
        let meanX = Double(n - 1) / 2.0
        let meanY = temps.reduce(0, +) / Double(n)
        let num = (0..<n).reduce(0.0) { $0 + (Double($1) - meanX) * (temps[$1] - meanY) }
        let den = (0..<n).reduce(0.0) { $0 + pow(Double($1) - meanX, 2) }
        guard den != 0 else { return "stable" }
        let slope = num / den
        if slope > 0.15 { return "increasing" }
        if slope < -0.15 { return "decreasing" }
        return "stable"
    }

    // MARK: - Footprint accounting

    var backendSizeMethod: String { "valkey:used_memory_dataset" }

    func backendSizeBytes() -> Int64 {
        guard let info = DazzleServer.shared.directCommand("INFO memory") else { return -1 }
        return parseValkeyUsedMemoryDataset(info)
    }

    func backendSizeBreakdown() -> [String: Int64]? {
        guard let info = DazzleServer.shared.directCommand("INFO memory") else { return nil }
        return parseValkeyMemoryStats(info)
    }
}
